mlir_graphblas/algorithms.py

```diff
@@ -2003,3 +2003,375 @@ def __call__(
+            irb.llvm.store(c_vector_ptr8, destination_ptr)
+
+        n_features_out = num_taus
+        n_features_out_i64 = num_taus_i64
+        heat_print_matrices = irb.llvm.alloca(n_features_out_i64, "!llvm.ptr<i8>")
+
+        twf_old = signal
+        twf_cur = irb.graphblas.matrix_multiply(L, signal, "plus_times")
+        twf_cur = irb.graphblas.apply(twf_cur, "div", right=a)
+        # TODO this should be handled by irb.graphblas.update
+        twf_cur = irb.graphblas.union(twf_cur, signal, "minus")
+
+        with irb.for_loop(0, n_features_out) as heat_print_for_vars:
+            idx = heat_print_for_vars.iter_var_index
+            idx_i64 = irb.arith.index_cast(idx, "i64")
+            c_vector_ptr = irb.llvm.getelementptr(c_vectors, idx_i64)
+            c_vector_ptr8 = irb.llvm.load(c_vector_ptr, "!llvm.ptr<i8>")
+            c_vector = irb.util.ptr8_to_tensor(c_vector_ptr8, "tensor<?xf64, #CV64>")
+            # assumes c_vector is dense
+            c_vector_values = irb.sparse_tensor.values(c_vector)
+            c_vector_0 = irb.memref.load(c_vector_values, c0)
+            c_vector_1 = irb.memref.load(c_vector_values, c1)
+            c_vector_0_over_2 = irb.arith.mulf(c_vector_0, c_0_5_f64)
+            heat_print_matrix = irb.graphblas.apply(
+                twf_old, "times", right=c_vector_0_over_2
+            )
+            right_summand = irb.graphblas.apply(twf_cur, "times", right=c_vector_1)
+            irb.graphblas.update(right_summand, heat_print_matrix, accumulate="plus")
+            heat_print_matrix_ptr8 = irb.util.tensor_to_ptr8(heat_print_matrix)
+            destination_ptr = irb.llvm.getelementptr(heat_print_matrices, idx_i64)
+            irb.llvm.store(heat_print_matrix_ptr8, destination_ptr)
+
+        a_diag = irb.graphblas.apply(eye, "second", right=a)
+        factor = irb.graphblas.union(L, a_diag, "minus")
+        two_over_a = irb.arith.divf(c2_f64, a)
+        # TODO this apply should be done in place
+        factor = irb.graphblas.apply(factor, "times", right=two_over_a)
+
+        twf_cur_ptr8_init = irb.util.tensor_to_ptr8(twf_cur)
+        twf_old_ptr8_init = irb.util.tensor_to_ptr8(twf_old)
+        twf_cur_ptr8 = irb.new_var("!llvm.ptr<i8>")
+        twf_old_ptr8 = irb.new_var("!llvm.ptr<i8>")
+        with irb.for_loop(
+            2,
+            N,
+            iter_vars=[
+                (twf_cur_ptr8, twf_cur_ptr8_init),
+                (twf_old_ptr8, twf_old_ptr8_init),
+            ],
+        ) as k_for_vars:
+            k = k_for_vars.iter_var_index
+            twf_cur = irb.util.ptr8_to_tensor(twf_cur_ptr8, "tensor<?x?xf64, #CSR64>")
+            twf_old = irb.util.ptr8_to_tensor(twf_old_ptr8, "tensor<?x?xf64, #CSR64>")
+
+            twf_new = irb.graphblas.matrix_multiply(factor, twf_cur, "plus_times")
+            # TODO this should be handled by irb.graphblas.update
+            twf_new = irb.graphblas.union(twf_new, twf_old, "minus")
```

Comment:
Same comment as above

Reply:
https://github.com/metagraph-dev/mlir-graphblas/commit/ce26f591efd6031e53380d12bb5641ce99a4d9bd implements this suggestion.